### PR TITLE
ENH: `Styler.apply` accept ndarray return with `axis=None` for consistency 

### LINKED ISSUE #39359: ENH: `Styler.apply` with `axis=None` can accept ndarray
Currently `Styler.apply` with `axis=0` and `axis=1` permit functions which return array-like objects, provided they are of appropriate shape. With `axis=None` the return *must* be a `DataFrame`.

This means the following example:

```
def highlight_max(d):
    """Highlight the maximum in a Series or DataFrame bold-orange."""
    css = 'background-color: orange; font-weight: bold;'
    return np.where(d == np.nanmax(d.values), css, None)

df.style.apply(highlight_max, axis=0)  # works
df.style.apply(highlight_max, axis=1)  # works
df.style.apply(highlight_max, axis=None)  # raises TypeError since return is not DataFrame
```

#### Describe the solution you'd like

The solution is a relatively simple change to code as part of the `Styler._apply` method that checks for `array-like` (or specifically `ndarray`) result instead of `DataFrame`, and constructs a `DataFrame` from that data if needs be, with appropriate index and cols.

#### API breaking implications

None since all previous behaviour is replicable.

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #39359
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Also minor edits to docs to extend information and show examples.

cc @TomAugspurger

### CODE REVIEW

**Comment by attack68** [comment]
this is ready for review.

It fixes the functionality and edits the docs where necessary, including some better examples.

**Review by jreback — CHANGES_REQUESTED** [review]
not sure we should change this. why can't the user just wrap the return value *inside* the apply.

File: doc/source/user_guide/style.ipynb [hunk 1]
```diff
@@ -368,11 +366,9 @@
...
    "source": [
     "def highlight_max(s):\n",
-    "    '''\n",
-    "    highlight the maximum in a Series yellow.\n",
-    "    '''\n",
-    "    is_max = s == s.max()\n",
-    "    return ['background-color: yellow' if v else '' for v in is_max]"
+    "    \"\"\"Highlight the maximum in a Series bold-orange.\"\"\"\n",
+    "    css = 'background-color: orange; font-weight: bold;'\n",
+    "    return np.where(s == np.nanmax(s.values), css, None)"
```
Comment (jreback) [inline doc/source/user_guide/style.ipynb:371]: you can simply use s.where no?

**Comment by attack68** [comment]
> not sure we should change this. why can't the user just wrap the return value _inside_ the apply.

Dont know why this restriction was placed in the first place. If you use `axis in [0,1]` then your function must return 1-dim array like (numpy array allowed), but `axis=None` must need a `DataFrame`. This makes it quite painful to implement simple numpy functions like `nanmax` cause the user has to deal with the different cases and match the return type.

The builtin functions always have, for example, logic like this:
```
        if data.ndim == 1:  # Series from .apply
            return [attr if v else "" for v in extrema]
        else:  # DataFrame from .tee
            return pd.DataFrame(
                np.where(extrema, attr, ""), index=data.index, columns=data.columns
            )
```

I  dont really understand the downside to just handling an array function output in all cases for axis.

File: doc/source/user_guide/style.ipynb (hunk 1 above)
Comment (attack68) [reply doc/source/user_guide/style.ipynb:371]: i'm using the values in `s` to create a new array either with `newval1` or `newval2` conditionally. I thought with `s.where` you can either keep `existingval` or `newval1` conditionally? Very possible im missing something here..

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -894,7 +900,7 @@ def apply(
             ``func`` should take a Series or DataFrame (depending
             on ``axis``), and return an object with the same shape.
             Must return a DataFrame with identical index and
-            column labels when ``axis=None``.
+            column labels or an ndarray of appropriate shape when ``axis=None``.
```
Comment (jreback) [inline pandas/io/formats/style.py:903]: same shape as input

Comment (jreback) [reply pandas/io/formats/style.py:903]: add versionchanged 1.3

**Review by jreback — CHANGES_REQUESTED** [review]
ok @attack68 this is ok, yeah consistency is better here. small comments, ping on green.

**Comment by attack68** [comment]
@jreback changes done. mostly green. fails unrelated it seems

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**